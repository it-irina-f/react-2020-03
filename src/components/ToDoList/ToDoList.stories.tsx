import React from "react";
import { withKnobs, number, array } from "@storybook/addon-knobs";

import { ToDoList } from "./ToDoList";

export default {
  title: "ToDoList Component",
  decorators: [withKnobs],
};

export const componentToDoList = () => <ToDoList />;
