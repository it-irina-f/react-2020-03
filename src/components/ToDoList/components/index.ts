export * from "./List";
export * from "./AddForm";
