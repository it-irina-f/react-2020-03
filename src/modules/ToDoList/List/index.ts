export * from "./List";
export * from "./ListItemEdit";
