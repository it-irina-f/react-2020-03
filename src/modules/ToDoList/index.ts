export * from "./ToDoList";
export * from "./AddForm";
export * from "./Filter";
export * from "./List";
export { todoSlice } from "./reducer";
export { getToDoModule } from "./module";
export { todoSaga } from "./saga";
