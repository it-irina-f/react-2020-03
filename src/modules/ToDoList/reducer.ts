import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { ToDoState } from "@/AppStore";

interface ListItemProps {
  id: number;
  text: string;
  isComplete: boolean;
}

export const selectors = {
  todo: ({ todo }: ToDoState) => todo,
};

export const initialState: {
  isLoading: boolean;
  editId: number;
  filter: string;
  list: ListItemProps[];
  queryResult: ListItemProps[];
  textInput: string;
} = {
  list: [],
  isLoading: false,
  editId: -1,
  filter: "all",
  queryResult: [],
  textInput: "",
};

export const todoSlice = createSlice({
  name: "todo",
  initialState,
  reducers: {
    setList: (state, { payload }: PayloadAction<any>) => {
      return {
        ...state,
        list: payload,
      };
    },
    addListItem: (state, { payload }: PayloadAction<string>) => {
      const newList = [
        ...state.list,
        {
          id: Date.now(),
          text: payload,
          isComplete: false,
        },
      ];
      return {
        ...state,
        list: newList,
      };
    },
    changeFilter: (state, { payload }: PayloadAction<string>) => {
      return {
        ...state,
        filter: payload,
      };
    },
    searchItem: (state, { payload }: PayloadAction<string>) => {
      const queryResult = state.list.filter((row) => {
        return row.text.toLowerCase().indexOf(payload.toLowerCase()) != -1;
      });

      return {
        ...state,
        queryResult: queryResult,
      };
    },
    toggleComplete: (state, { payload }: PayloadAction<number>) => {
      const updList = state.list.map((row) => {
        if (row.id === payload) {
          return { ...row, isComplete: !row.isComplete };
        }
        return row;
      });
      return {
        ...state,
        list: updList,
      };
    },
    deleteItem: (state, { payload }: PayloadAction<number>) => {
      const updList = state.list.filter((row) => {
        return row.id !== payload;
      });
      return {
        ...state,
        list: updList,
      };
    },
    editItem: (state, { payload }: PayloadAction<number>) => {
      const textInput =
        payload != -1
          ? state.list.filter((row) => {
              return row.id === payload;
            })[0].text
          : "";
      return {
        ...state,
        editId: payload,
        textInput: textInput,
      };
    },
    saveItem: (state, { payload }: PayloadAction<string>) => {
      const updList = state.list.map((row) => {
        if (row.id === state.editId) {
          return { ...row, text: payload };
        }
        return row;
      });

      return {
        ...state,
        list: updList,
        editId: -1,
        textInput: "",
      };
    },
  },
});

export const { reducer, actions } = todoSlice;
