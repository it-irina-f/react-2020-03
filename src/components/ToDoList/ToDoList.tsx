import React from "react";
import type { ListItemProps, ToggleCompleteProps } from "types/todo";
import { List } from "./components/List";
import { AddForm } from "./components/AddForm";
import styled from "@emotion/styled";

interface ToDoListProps {
  list: ListItemProps[];
}

interface ToDoListState {
  list: ListItemProps[];
}

const ToDoListWrapper = styled.div`
  display: block;
  width: 500px;
  margin: 10px auto;
  padding: 20px 50px;
  background: #eaeaea;
`;

const TitleWrapper = styled.div`
  padding: 10px;
  text-transform: uppercase;
  font-weight: bold;
  font-size: 18px;
  text-align: center;
`;

export class ToDoList extends React.Component<ToDoListProps, ToDoListState> {
  constructor(props: ToDoListProps) {
    super(props);
    this.state = {
      list: props.list,
    };
    this.toggleCompleteHandler = this.toggleCompleteHandler.bind(this);
  }

  public toggleCompleteHandler(id: number): void {
    const updateList = this.state.list.map((row) => {
      if (row.id === id) {
        return { ...row, isComplete: !row.isComplete };
      }
      return row;
    });

    this.setState({
      list: updateList,
    });
  }

  render() {
    return (
      <ToDoListWrapper>
        <TitleWrapper>Список дел</TitleWrapper>
        <List
          list={this.state.list}
          toggleComplete={this.toggleCompleteHandler}
        />
        <AddForm />
      </ToDoListWrapper>
    );
  }
}
