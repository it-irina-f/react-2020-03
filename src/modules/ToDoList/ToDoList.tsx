import React from "react";
import { List } from "@/modules/ToDoList/List";
import { AddForm } from "@/modules/ToDoList/AddForm";
import { Filter } from "@/modules/ToDoList/Filter";
import { Search } from "@/modules/ToDoList/Search";
import styled from "@emotion/styled";
import { reactLocalStorage } from "reactjs-localstorage";

interface ListItemProps {
  id: number;
  text: string;
  isComplete: boolean;
}

interface ToDoListProps {
  list?: ListItemProps[];
}

interface ToDoListState {
  list: ListItemProps[];
  isLoading: boolean;
  editId: number;
  filter: string;
  queryResult: ListItemProps[];
}

const ToDoListWrapper = styled.div`
  display: block;
  width: 800px;
  margin: 10px auto;
  padding: 20px 50px;
`;

const TitleWrapper = styled.div`
  padding: 10px;
  text-transform: uppercase;
  font-weight: bold;
  font-size: 18px;
  text-align: center;
`;

const ManageWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  margin: 20px 0;
`;

export class ToDoList extends React.Component<ToDoListProps, ToDoListState> {
  constructor(props: ToDoListProps) {
    super(props);
    this.state = {
      list: [],
      isLoading: true,
      editId: -1,
      filter: "all",
      queryResult: [],
    };
  }

  componentDidMount(): void {
    this.timerHandle();
  }

  componentWillUnmount(): void {
    clearTimeout(this.timerHandle());
  }

  timerHandle = () => setTimeout(() => this.getList(), 3000);

  getList(): void {
    const todosFromLocalStorage = reactLocalStorage.getObject(
      "localStorageTodos"
    );
    this.setState({
      isLoading: false,
      list: Array.isArray(todosFromLocalStorage) ? todosFromLocalStorage : [],
    });
  }

  updateList(updateList): void {
    this.setState(
      {
        list: updateList,
      },
      () => {
        reactLocalStorage.setObject("localStorageTodos", this.state.list);
      }
    );
  }

  toggleCompleteHandler = (id: number) => {
    const updList = this.state.list.map((row) => {
      if (row.id === id) {
        return { ...row, isComplete: !row.isComplete };
      }
      return row;
    });

    this.updateList(updList);
  };

  addListItemHandler = (text: string) => {
    const newList = [
      ...this.state.list,
      {
        id: Date.now(),
        text: text,
        isComplete: false,
      },
    ];

    this.updateList(newList);
  };

  deleteItemHandler = (id: number) => {
    const updList = this.state.list.filter((row) => {
      return row.id !== id;
    });

    this.updateList(updList);
  };

  editHandler = (id: number) => {
    this.setState({
      editId: id,
    });
  };

  saveItemHandler = (id: number, text: string) => {
    const updList = this.state.list.map((row) => {
      if (row.id === id) {
        return { ...row, text };
      }
      return row;
    });

    this.updateList(updList);

    this.editHandler(-1);
  };

  changeFilterHandler = (mode: string) => {
    this.setState({
      filter: mode,
    });
  };

  searchItemHandler = (text: string) => {
    const queryResult = this.state.list.filter((row) => {
      return row.text.toLowerCase().indexOf(text.toLowerCase()) != -1;
    });

    this.setState({
      queryResult: queryResult,
    });
  };

  render() {
    return (
      <ToDoListWrapper>
        <TitleWrapper>Список дел</TitleWrapper>
        <ManageWrapper>
          <Search searchItem={this.searchItemHandler} />
          <Filter changeFilter={this.changeFilterHandler} />
        </ManageWrapper>
        {this.state.isLoading ? (
          <h1>Загрузка данных...</h1>
        ) : (
          <List
            list={
              this.state.queryResult.length === 0
                ? this.state.list
                : this.state.queryResult
            }
            editId={this.state.editId}
            toggleComplete={this.toggleCompleteHandler}
            deleteListItem={this.deleteItemHandler}
            handleEdit={this.editHandler}
            saveListItem={this.saveItemHandler}
            filter={this.state.filter}
          />
        )}
        <AddForm addListItem={this.addListItemHandler} />
      </ToDoListWrapper>
    );
  }
}
