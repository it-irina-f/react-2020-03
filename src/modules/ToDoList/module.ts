import { ISagaModule } from "redux-dynamic-modules-saga";

import { reducer } from "./reducer";
import { todoSaga } from "./saga";

export const getToDoModule = (): ISagaModule<typeof reducer> => ({
  id: "todo",
  reducerMap: {
    todo: reducer,
  },
  sagas: [todoSaga],
});
