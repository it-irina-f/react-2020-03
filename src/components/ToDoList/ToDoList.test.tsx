import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";

import { ToDoList } from "./ToDoList";

const click = jest.fn();

const list = [
  { id: 0, text: "Полить цветы", isComplete: true },
  { id: 1, text: "Сделать ДЗ", isComplete: false },
  { id: 2, text: "Купить продукты", isComplete: true },
  { id: 3, text: "Приготовить ужин", isComplete: false },
  { id: 4, text: "Уборка квартиры", isComplete: false },
];

describe("Component ToDoList", () => {
  it("renders list", () => {
    expect(
      renderer.create(<ToDoList list={list} />).toJSON()
    ).toMatchSnapshot();
  });
});
