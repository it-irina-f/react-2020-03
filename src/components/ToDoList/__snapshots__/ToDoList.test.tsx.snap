// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component ToDoList renders list 1`] = `
<div
  className="css-vtf7xb"
>
  <div
    className="css-k87agx"
  >
    Список дел
  </div>
  <h1>
    Загрузка данных...
  </h1>
  <form>
    <input
      name="addlistItem"
      placeholder="Добавить новую задачу"
      type="text"
    />
    <button
      onClick={[Function]}
      type="button"
    >
      Добавить
    </button>
  </form>
</div>
`;
